src/gui/protocolitemmodel.cpp

```diff
@@ -0,0 +1,212 @@
+                || status == SyncFileItem::DetailError
+                || status == SyncFileItem::BlacklistedError) {
+                return Theme::instance()->syncStateIcon(SyncResult::Error);
+            } else if (Progress::isWarningKind(status)) {
+                return Theme::instance()->syncStateIcon(SyncResult::Problem);
+            } else {
+                return {};
+                // TODO: display icon on success?
+                //                return Theme::instance()->syncStateIcon(SyncResult::Success);
+            }
+        }
+    case Models::UnderlyingDataRole:
+        switch (column) {
+        case ProtocolItemRole::Time:
+            return item.timestamp();
+        case ProtocolItemRole::Folder:
+            return item.folderName();
+        case ProtocolItemRole::Action:
+            return item.message();
+        case ProtocolItemRole::Size:
+            return item.size();
+        case ProtocolItemRole::File:
+            return item.path();
+        case ProtocolItemRole::Account:
+            return getFolder(item)->accountState()->account()->displayName();
+        case ProtocolItemRole::Count:
+            Q_UNREACHABLE();
+        }
+    }
+    return {};
+}
+
+QVariant ProtocolItemModel::headerData(int section, Qt::Orientation orientation, int role) const
+{
+    if (orientation == Qt::Horizontal) {
+        const auto actionRole = static_cast<ProtocolItemRole>(section);
+        switch (role) {
+        case Qt::DisplayRole:
+            switch (actionRole) {
+            case ProtocolItemRole::Time:
+                return tr("Time");
+            case ProtocolItemRole::File:
+                return tr("File");
+            case ProtocolItemRole::Folder:
+                return tr("Folder");
+            case ProtocolItemRole::Action:
+                return _issueMode ? tr("Issues") : tr("Action");
+            case ProtocolItemRole::Size:
+                return tr("Size");
+            case ProtocolItemRole::Account:
+                return tr("Account");
+            case ProtocolItemRole::Count:
+                Q_UNREACHABLE();
+            };
+
+        case Models::StringFormatWidthRole:
+            switch (actionRole) {
+            case ProtocolItemRole::Time:
+                return 20;
+            case ProtocolItemRole::Folder:
+                return 30;
+            case ProtocolItemRole::Action:
+                return 15;
+            case ProtocolItemRole::Size:
+                return 6;
+            case ProtocolItemRole::File:
+                return 64;
+            case ProtocolItemRole::Account:
+                return 20;
+            case ProtocolItemRole::Count:
+                Q_UNREACHABLE();
+            }
+        };
+    }
+    return QAbstractTableModel::headerData(section, orientation, role);
+}
+
+void ProtocolItemModel::addProtocolItem(const ProtocolItem &&item)
+{
+    Q_ASSERT(actualSize() == _data.size());
```

Comment:
The logic within this function doesn't quite get clear to me. It's very confusing. I understand why you call these `begin`/`end` callbacks. However, it seems a bit inconsistent to me.

I am furthermore pretty sure you should replace `_data` with a better suited structure that encapsulates the logic of keeping track of the current and max sizes. Makes the entire thing easier to use and test. How about a ring buffer class that implements an `append()` operation...? You don't have to care about the underlying data structure then, too.

You could implement the `begin*`/`end*` stuff with signals/slots.

In fact, an `std::deque` would work a lot better, too. It already features an `append`, and `pop`ing from the front gets rid of the first element.

Edit: oh, now I see. You want to keep all the previous entries, right? That's why you use `_start` to keep track of the start.
In fact, this reinforces my opinion that the data structure that stores the logs should be separated. I think you want a class that wraps an `std::deque` (because of its fast appending) and implements the logic of `_start` and `_maxLogSize` (i.e., basically a view on said `deque`), which is used as a data structure in here...

Edit 2: strike that. It's a ring buffer implemented in a vector, `_start` is just used to keep track of the current position for new inserts.

Reply:
Efficiency wise, an `std::deque` might be sufficient to implement your logic. But a dedicated ring buffer interface is surely nicer to use, and the underlying data structure is hidden behind abstraction. Any kind of performance improvement can be done later on.

Reply:
Here's a basic implementation based on an `std::deque`: https://gist.github.com/fmoc/9b52f831c11835f93276b301211dcd8e